### ISSUE TITLE
Separate API server

### ISSUE THREAD

**Issue by rzinkstok** [comic/grand-challenge.org#631]
We think it would make sense to decouple the API server from the current server.

Advantages/thoughts
- the settings file for the API does not need any of the frontend stuff (templates, CSRF, messages, clickjacking, summer note, bleach, …) and even timezone stuff etc should not be needed (all relegated to the API consumer)
- The API server and the current server can run happily alongside each other, sharing the same database and celery infrastructure
- the API server can be put an a separate docker container
- when going to a complete new frontend (e.g. react, which we are considering) the API server will become the backend server, enabling a smooth transition and/or the option to co-exist with the current frontend.

The biggest change will be in the Django settings file: we think the differences are too big to do this with if/else statements in a single settings file, so we propose splitting the current settings file into two separate settings file, one for the current server and one for the API server. Of course this has drawbacks, so perhaps we can think of a better way to handle the settings.

We have started implementing this in the eyra-api-dev branch, see https://github.com/comic/grand-challenge.org/tree/eyra-api-dev.

We'd be happy to hear your thoughts!

**Comment by jmsmkn** [comment]
I'm struggling to see the benefits of this. The unnecessary code for the API is hardly an overhead compared to the rest of the app (both in footprint and I predict response time for API endpoints too) and you run the risk of divergence - a lot of the stuff you mention is actually necessary. Summer note has all the tables for user uploaded media, and bleach is used to sanitise input serverside.

**Comment by rzinkstok** [comment]
The main issue we were struggling with involved authentication of the rest api: we want to be able to do authentication through the API alone, both for normal users and for social-auth users. But your points are valid of course (as always!), so we looked more into the social-auth/rest-framework combination. We've managed to change it such that it works nicely together with the full grand-challenge server, so we don't need a separate settings.py anymore. 

We've set it up such that a post to /api/login/ with valid credential logs a non-social user in, and a get to /api/social/login/google-oauth2/ starts social authentication. Either way, after successful authentication you end up with a Django authentication token, to be used in subsequent calls.

See https://github.com/comic/grand-challenge.org/tree/eyra-api-dev for the current state. Is this better in your opinion?
